### DISCUSSION TITLE
LEDs not working

### DISCUSSION THREAD

**Discussion by MrPresident2** [awawa-dev/HyperHDR#203] [Q&A]
Hey guys,

I don't know what to do and I really need your help.

I got the Pi zero 2w and when I connect the LEDs to GPIO 18, nothing happens. I tried with 2 different LEDs types ( WS2812 and SK6812) and nothing is working.

I tried connecting it to a different GPIO (19) but nothing works ( I can only change the GPIO under WS2812 strip).

I tried HyperHDR 16 and 17 and regular Hyperion nothing works.

Could it be a faulty pi?

any suggestions?

Hope you can help me :)

**Comment by awawa-dev** [comment]
1) Driving non-SPI LED strip directly from Rpi is not officially supported or recommended by HyperHDR, but possible for advanced users. Definitely it's not a way for everyone and always it posses also a risk for Raspberry Pi.  
2) Assuming your Rpi GPIO is not fried, you could use this settings: link https://github.com/awawa-dev/HyperHDR/discussions/111 . They are well tested and definitely works fine when everything is OK.  
3) But for doing that root permissions are required. Link: https://github.com/awawa-dev/HyperHDR/issues/52 Verify your HyperHDR logs if there are red errors while initializing LED device (*mmap() failed* error).  
4) The problem can be also have an electrical nature. Does your setup (LED and Raspberry Pi) have a common ground? Voltage level shifter 3.3V to 5V is present?

![obraz](https://user-images.githubusercontent.com/69086569/151852166-d9b42f28-5385-4c01-a170-e2e8f4af7c7e.png)

**Comment by awawa-dev** [comment]
And  Pi zero 2w platform is supported only by v18beta. You can find installers in latest https://github.com/awawa-dev/HyperHDR/actions
artifacts. That one of the other reasons why that method is not officially supported: ws281x library needs to be updated when new board or revision is released. 
-  https://github.com/awawa-dev/HyperHDR/discussions/147

**Comment by MrPresident2** [comment]
Thank you for your response. I am using the SK6812, not the WS281x. So what I'm understanding is the only way to make it work is via ESP board with WLED? I'm trying to do that but I'm having a huge delay and I can only make it work with about 3 fps which makes it looks really bad.  in all of the vids I saw on youtube people connected the LED directly to the PI. 
Now, I'm trying to find the latest beta version, 18 but I can't find anywhere to download it.

**Comment by awawa-dev** [comment]
No, HyperSerialEsp8266/HyperSerialEsp32/HyperSerialWLED or recommended for Rpi: HyperSPI are cable solutions so extremely reliable and fast. You can find more details on their project homepages.

SK6812 is using same WS281x library (SPI driver version never worked for me so can't recommend it). You have it on the previous link with example configuration https://github.com/awawa-dev/HyperHDR/discussions/111

v18 beta installers are in the latest action tab. So for today that's one:
![obraz](https://user-images.githubusercontent.com/69086569/151862784-22041ded-9f74-442b-adce-19c2d30032ac.png)

**Comment by MrPresident2** [comment]
First of all, I really appreciate your help. Thank you very much. I've been trying to make it work for weeks now.
Can you explain or maybe an article about how to make the wired connection between the esp and the pi?

**Comment by awawa-dev** [comment]
Did you look for these projects? For HyperSPI there are even screenshots how to connect ESP to Rpi: https://github.com/awawa-dev/HyperSPI works with every generic ESP.


For HyperSerialEsp8266/HyperSerialEsp32/HyperSerialWLED you need first determine using your eyes (search for chip markings, they are barely visible) which serial port chip you have on your ESP board. Only fast 2Mb speed chips like CH340G/C, CH9102x or CP2104 are supported.

Connection is very simple: USB.

**Comment by MrPresident2** [comment]
Thank you again, I have Esp8266 CH340G, I installed the HyperSerialWLED and you say I need to connect them both via USB? I only have one USB on my pi zero and it is connected to the capture card

**Comment by awawa-dev** [comment]
CH340G is good. 
So possible solution for now:

1) Lack of free USB port could be a problem for HyperSerialEsp8266/HyperSerialEsp32/HyperSerialWLED and you can solve it with USB hub. But it will complicate your setup.

2) Use HyperSPI (you need 3 SPI connection from RPI GPIO to ESP8266 GPIO and additional one that connects both RPI&ESP8266 GPIO grounds).

3) you can try to install v18 beta and use Rpi directly with sk6812. with v16 or v17 almost for sure you have errors in HyperHDR log since RPi zero 2 W were not supported then (verify it in logs). Since you are almost there you can give it a chance first. Still, common ground between LED strip ground and Raspberry Pi is necessary if each uses different power supply (if you are not sure how to connect it you should stop here and look for diagrams).

**Comment by MrPresident2** [comment]
Again, I really appreciate your help.
Still running HyperHDR V17
-So I tried it with HyperSPI and the LEDs and the boards are connected with the same power supply. nothing works
-Now, I'm trying it with HyperSerialWLED and the ESP is connected with USB (I'll try it without the capture card just to see if the Pi controls the ESP) the ESP works but it seems like nothing is controlling it, I'm getting the regular WLED page not the one that says controlled by something else.
-Now just to be sure, when I'm using HyperSerialWLED what setting do I choose in the controller type at HyperHDR? SK6812spi? or something else?
-Now installing the HDR v18, I see there DEB(30 MB) files and not Image(500MB like in the not beta version), so I'm not sure how to install that version

I know I might be annoying nob, but that's the only way to learn is by asking questions. I REALLY appreciate your help.
Thank you, sir.

**Comment by awawa-dev** [comment]
You are too fast, jumping from one solution to another without deeper diagnostic ;)
1) HyperSPI has debug output on the serial port which could tell us many useful information about the cause of the problem. Wrong connection? Wrong pinout? OK skip it for now.
2) HyperSerialWLED: did you configure HyperHDR using 2000000 speed and enabled AWA protocol in HyperHDR adalight? That driver (and proper hardware: CH340G) you need to make HyperSerialWLED working. ![](https://user-images.githubusercontent.com/69086569/130987180-f12fb88e-850a-4fc3-8ea8-3579d900059a.png)
3) Did you verify using your EYES if you have CH340G or that's how the product was just advertised in the shop? :) There are many counterfeits.

**Comment by awawa-dev** [comment]
make sure to select proper USB serial port in the 'output path'. Your CH340G should be visible in the list there.

**Comment by MrPresident2** [comment]
(still using V17 cause I don't really know how to install the 18beta)
Ok so after doing everything you said (btw yes I saw with my eyes the ESP has CH340G chip), now it works and the pi controls the WLED via USB. so that is great!!! the next thing is to try and see something because I don't have the capture card connected, so I downloaded Android Hyperion Grabber, welllllll... my tv is dying, youtube gets 5 frames top haha and plex gets maybe 1fps, so I don't know about the app (maybe its a setting thingy but IDK. is there a setting to control the grabber app in HyperHDR like the setting for the capture card?)
Anyway, if I'm getting a USB hub, is that just a plug-and-play thing for the pi? because maybe I can use the capture card and the WLED at the same time, what do you think?

**Comment by awawa-dev** [comment]
Great! In my case it worked when USB grabber and ESP8266/ESP32 were connected through USB hub (and USB hub was transparent for my setup). But you must have SOLID 2A USB power supply for Raspberry Pi AND solid USB power cable to provide the current from the power supply to the Raspberry Pi. Same power supply for Raspberry Pi and LED strip is not a good prognostic at all: Raspberry Pi should always have dedicated power supply. Sooner or later you may end up in troubles.

**Comment by MrPresident2** [comment]
Everything is connected through the 5v 60A power supply, do you think I need another power supply for the Pi? I am having some weird problems that the capture card just disconnects and I need to restart the Pi and it happens a lot.
so the way that everything is connected: 
Pi connects to the power supply via GPIO (5v and ground)
ESP gets power through the USB from the Pi, the same as the Capture card (both connected through the HUB)
LEDs connected directly to the power supply + power injunction because I had some voltage drop in the white color
What do you think?

**Comment by awawa-dev** [comment]
It's not about max amperage: 60A should be fine for your setup. But such LED power supplies are stable enough for LED strip, not for raspberry pi. Especially when scene changes from dark to bright the load could from jump few percent to over 50-100% depending on the setup. The voltage won't be stable for Raspberry Pi, it only works stable in range 4.75 - 5.25V.I have never managed to provide such a stable system all the time when I use a shared power supply. Under-voltage can be detected in dmesg log (search for 'under-voltage detected' messages). Over-voltage could trigger self-resetting fuse if Raspberry Pi is powered by USB (when it is powered by GPIO it will get damaged since it's not protected).

**Comment by MrPresident2** [comment]
ok, I ordered a cable I'm missing to connect the pi to a different power supply, for now, I connected the Pi to the main power supply instead of the GPIO I connected it via the USB power. And that problem seems to disappear now. And the HomeAssistant integration works amazing. I can't stop looking at the TV :)
Thank you! and if the problem or something else will stop working haha ill let you know :)

**Comment by MrPresident2** [comment]
Ok talked too soon, the capture card still disconnects and disappears from the Hyperion, it comes back after reboot buttt I don't think rebooting the pi every 5-10 minutes is a solution. Any Idea why is it happening?

**Comment by awawa-dev** [comment]
As I said: verify ```dmesg``` & application log to see what's going on in the system when grabber disconnects. If you are using still the same power supply connected through USB won't change much. And after you get second power supply remember to provide common ground between Rpi/ESP and LED strip.

**Comment by MrPresident2** [comment]
I added a pic with the errors.
![Screenshot 2022-01-31 165619](https://user-images.githubusercontent.com/60242667/151897290-88a139f8-2209-4d91-b7ae-9eb149d2f647.jpg)

I don't see anything about voltage errors.

And do you have a pic of what common ground is? I'm not sure I understand it completely.

**Comment by awawa-dev** [comment]
1 Did you something to the video signal when that happen? Stopping the video, changing the video resolution that could break video streams that grabber were capturing?
2 ssh to the raspberry using for example putty and execute ```dmesg```. Verify & post the result output.

**Comment by MrPresident2** [comment]
No, just the Tv screen saver, I was waiting for it to happen again.

I added the logs from putty
![Log2](https://user-images.githubusercontent.com/60242667/151898364-f38cbea8-0e67-48f4-961e-c7a819e8d07a.jpg)


Could it be because I'm using V17 and not 18beta?

**Comment by awawa-dev** [comment]
Your grabber disconnected itself and then reconnected after 0.5s. That's bad because there is no obvious reason and MS2109 should not do it. Maybe it's power supply issue or maybe it's already damaged.

OK, there is workaround (but not designed for such behavior): enable 'auto-resume' in the grabber properties.

**Comment by MrPresident2** [comment]
I'm using a capture card not the grabber app, I found a USB hub so could it be that the capture card is faulty, or is it the Pi?
I will try the auto-resume and hope it will work because when the capture card disconnect it does not show anymore in HyperHDR UI and it will only reappear after Sudo reboot

**Comment by awawa-dev** [comment]
Yes, you are using MS2109 It's generic name of that grabber family ;) The problem is hardware because the grabber is disconnecting from the system and we can only hope that it will reconnect in 2seconds. Then we try to activate it again. That's how the auto-resume option works. Unfortunately such pause will be visible on your LED strip. 
Some grabbers can behave like that when they lose HDMI source and that's why I asked you about it. But never met it with MS2109.

**Comment by MrPresident2** [comment]
OK, I see, it happened again and I added the log here.
So if it is the grabber, ill try and find another one, and ill see if it happen again.
btw the auto-resume didn't work.
![log3](https://user-images.githubusercontent.com/60242667/151899659-866f9b8e-1108-4d0d-95ec-8188cb476a75.jpg)

**Comment by awawa-dev** [comment]
Yes, in this case the break was 6 seconds...too long to give auto-resume a chance (~2 second limit).
Still not sure if the problem is the grabber or the power supply. Both can cause such problems.

**Comment by MrPresident2** [comment]
Can you explain to me about the common ground so ill do it the right way?
And I will in a few days when ill get a different capture card and il let you know if it solved the problem

**Comment by awawa-dev** [comment]
Connect the ground from ESP8266 GPIO (or Raspberry Pi GPIO if ESP8266 is connected to its USB port: in this case they share it too) to the ground of LED strip input (or to the same line that is connected to LED strip ground).

**Comment by MrPresident2** [comment]
So I tried it with my other... capture card... its overkill and not for that but that's what I have here (Elgato capture card)
still the same, these are the logs. 
![log4](https://user-images.githubusercontent.com/60242667/151902006-3097c0b7-c219-427b-af04-6d7776fe30b1.jpg)


So I guess the only thing left to do is the power supply?

**Comment by awawa-dev** [comment]
To be precise: I do not see the disconnection itself here, only a UVC kernel driver error, but I assume that it happened later. I would focus on the basic setup to stabilize it first: just Raspberry Pi, USB grabber and ordinary USB power supply. Disconnect and do not use LED strip, ESP8266 and LED power supply for now. Do you use some kind of cable or adapter for Raspberry Pi to connect a grabber? Maybe it's faulty? You have to test everything left to confirm LED strip power supply as a cause of the problem and to rule out the worst case scenario (Raspberry pi).

**Comment by MrPresident2** [comment]
I really appreciate your help. 
I do have an adapter from micro USB from the pi to regular type a from there I connect the HUB.
I order different cables and adapters to debug this if not I will connect a different power supply and do what you said with no LEDs or ESP.
It will probably won't be today(since I just ordered everything), so I will let you know when I get the parts and do it.
Again, Thank you so so much for the help.

**Reply by corckie** [reply]
I'm using WS2815 with GPIO just fine, I know it's not officially supported, but for me it is convenient and have no problems whatsoever. 

To make it work: 

in config.txt put the line dtparam=audio=off (take sd card and edit config file from your PC)

Insert card

Use terminal or SSH for these commands:
sudo systemctl disable --now hyperhdr@pi
sudo systemctl enable --now hyperhdr@root

Restart your LED power supply, restart RPI.

Should work from now on.

**Comment by MrPresident2** [comment]
@awawa-dev  Update:
I got all the new cables+new capture card.
now, everything is working as it should. Thank you for your help!
Now I have another question, I think I need to calibrate my colors. I'm trying to use the calibration feature in hyperhdr but it asks for pictures and other stuff, do you know how can I do the calibration or where can I find those stuff?

**Reply by awawa-dev** [reply]
no. but can use or prepare in the paint app boards/pictures with solid colors (red/blue/green/cyan/gray etc) or youtube [for example: Monitor Color Test / Monitor-Farbtest (RGB/CMYK) (1080p)](https://www.youtube.com/watch?v=JeQuelXsUYA)
